Vp8_enc: Implement VP8 SVC-T encoding

wudping opened the pull request:
Add an interface IVaapiFlagParameter that contains all the common functions for non-SVC-T and SVC-T.
Class VaapiFlagParameterNormal to implement picture parameters setting in non-SVC-T mode.
Class VaapiFlagParameterSVCT in SVC-T mode.

wudping commented:
please help to review. Thanks.

uartie commented:
Please fix build error

wudping commented:
Hi @uartie , 
The reason for 9495510ef698 compilation failure, is that the commit applies a variable "temporal_id” that still hasn't been merged into current Libva Master branch. My local Libva has the variable, so it can pass through the compilation. 

Error info:
dec::VaapiEncoderVP8::fill(VAEncPictureParameterBufferVP8*, const PicturePtr&, const SurfacePtr&) const':
13:03:48,452 INFO  - vaapiencoder_vp8.cpp:497:30: error: 'struct _VAEncPictureParameterBufferVP8::<anonymous union>::<anonymous>**' has no member named 'temporal_id'**
13:03:48,452 INFO  - **picParam->ref_flags.bits.temporal_id** = picture->m_temporalLayerID;

lizhong1008 commented on encoder/vaapiencoder_vp8.cpp:
```diff
@@ -49,14 +50,303 @@ class VaapiEncPictureVP8 : public VaapiEncPicture
-VaapiEncoderVP8::VaapiEncoderVP8():
-	m_frameCount(0),
-	m_qIndex(VP8_DEFAULT_QP)
+class IVaapiFlagParameter {
```
What does flag parameter mean? Do you mean FillParameter?

wudping replied:
It means the flags in the parameter. I can't figure out a proper name for that.

lizhong1008 commented on encoder/vaapiencoder_vp8.cpp:
```diff
@@ -49,14 +50,303 @@ class VaapiEncPictureVP8 : public VaapiEncPicture
+    Uint32Vector m_tempLayerIDs;
+    uint8_t m_layerNum;
+};
+
```
Why put the class declaration in the cpp file instead of hpp file?

wudping replied:
Because it just for vp8 encode, not for vp9 or any other codecs.

lizhong1008 replied:
you can move it to vaapiencoder_vp8.h

xuguangxin replied:
Hi @lizhong1008 ,
The class only used by this .cpp, it's no need add this to .h.

lizhong1008 commented on encoder/vaapiencoder_vp8.cpp:
```diff
@@ -49,14 +50,303 @@ class VaapiEncPictureVP8 : public VaapiEncPicture
+            printf(" : ");
+        printf("%d", m_framerateRatio[i]);
+    }
+    printf("\n");
```
Please change it to DEBUG instead of print

lizhong1008 replied:
All print() functions should be replaced to DEBUG()

lizhong1008 commented on interface/VideoEncoderDefs.h:
```diff
@@ -152,6 +153,11 @@ typedef struct VideoFrameRate {
 
+typedef struct SVCTVideoFrameRate {
+    VideoFrameRate fraction[32];
+    uint8_t num;
```
Here is the interface to user, would better add some comment to describe what does fraction and num mean.

xuguangxin replied:
Hi @wudping , why we need this?, why h264 not needed it?

uartie commented:
@wudping please provide a link to the libva PR that this depends on. thanks.

wudping commented:
hi, @uartie , 
The Libva code is on the Teamforge:
**url:** ssh://dwux@git-amr-3.devtools.intel.com:29418/otc_media-libva.git
**branch:** vp8-svct

lizhong1008 commented:
@wudping please make sure "picParam->ref_flags.bits.temporal_id" is necessary for vp8 svc-t decoder and VAAPI driver or not. If it is not necessary, you can remove the line "picParam->ref_flags.bits.temporal_id = picture->m_temporalLayerID;". Then it can be buildable using upstream VAAPI.

lizhong1008 commented:
@wudping ,if it is necessary, you can send a pull request to 01.org/libva.

xuguangxin commented:
@wudping will take some vacation,  I will help address the issue and refact the code

xuguangxin committed:
interface: add VideoTemproalLayers for temporal encoder

we will use this for h264 and vp8 temporal encoder, hope it can apply to
all svct encoders.

xuguangxin committed:
refact: h264, move temporal bitrate contorl to base

so we can remove duplicate code in h264 and vp8 svct encoder

xuguangxin committed:
interface: remove h264 specific temporal contorl fields

all fields move to VideoTemporalLayers

xuguangxin committed:
refact: isolate reference mannager code to Vp8Encoder class

this is preparation for vp8 svct encoder

Signed-off-by: wudping <dongpingx.wu@intel.com>
Signed-off-by: Xu Guangxin <guangxin.xu@intel.com>

xuguangxin committed:
vp8enc: add svct encoder

the getLayerIds logical is complicate, we can revisit it later

Signed-off-by: wudping <dongpingx.wu@intel.com>
Signed-off-by: Xu Guangxin <guangxin.xu@intel.com>

xuguangxin committed:
interface: add temporalID to VideoEncOutputBuffer

user can use this field to select output dest.
frames in diffrent layer may put in diffrent rtp packet

Signed-off-by: wudping <dongpingx.wu@intel.com>
Signed-off-by: Xu Guangxin <guangxin.xu@intel.com>

xuguangxin committed:
baseenc: set temporalID to output buffer

Signed-off-by: wudping <dongpingx.wu@intel.com>
Signed-off-by: Xu Guangxin <guangxin.xu@intel.com>

xuguangxin commented:
I am still working on this, I hope we can refact h264 code make the svct share same code with vp8 svct. Hope I can send h264 refact code in following days

wudping commented:
compile error as following is due to the libva master dose not contain the variable:

vaapiencoder_vp8.cpp:311:34: error: ‘struct _VAEncPictureParameterBufferVP8::<anonymous union>::<anonymous>’ has no member named ‘first_ref’
         **picParam->ref_flags.bits.first_ref = 0x01;
         picParam->ref_flags.bits.second_ref = 0x02;**
                                  ^
...
vaapiencoder_vp8.cpp:335:30: error: ‘struct _VAEncPictureParameterBufferVP8::<anonymous union>::<anonymous>’ has no member named ‘temporal_id’
     **picParam->ref_flags.bits.temporal_id = picture->m_temporalID;**

xuguangxin commented:
could you protect this with some libva version check?

```
compile error as following is due to the libva master dose not contain the variable:

vaapiencoder_vp8.cpp:311:34: error: ‘struct _VAEncPictureParameterBufferVP8::::’ has no member named ‘first_ref’
picParam->ref_flags.bits.first_ref = 0x01;
picParam->ref_flags.bits.second_ref = 0x02;
^
...
vaapiencoder_vp8.cpp:335:30: error: ‘struct _VAEncPictureParameterBufferVP8::::’ has no member named ‘temporal_id’
picParam->ref_flags.bits.temporal_id = picture->m_temporalID;

```

dongpingx-wu committed:
vp8enc: modify the layerID setting

According to libva, layer_id[] specifies the layer ids
for frames starting from the 2nd frame.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
vp8_enc: to fix layer frame referencing bug

To prevent layer 2nd's frames from referencing to the layer3's on SVC-T.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

dongpingx-wu committed:
vp8_svct: add compilation option for SVC-T

Adding a compilation option "--enable-vp8svct" to make sure that no problem
with the compilation, despite of driver unsupporting it.

Signed-off-by: dongping wu <dongpingx.wu@intel.com>

wudping commented:
updated, thanks.
